client/rpcserver/rpcserver.go

```diff
@@ -1,20 +1,403 @@
+	w.WriteHeader(code)
+	encoder := json.NewEncoder(w)
+	if err := encoder.Encode(thing); err != nil {
+		log.Errorf("JSON encode error: %v", err)
+	}
+}
+
+// handleJSON handles all https json requests.
+func (s *RPCServer) handleJSON(w http.ResponseWriter, r *http.Request) {
+	// All http routes are available over websocket too, so do not support
+	// persistent http connections. Inform the user and close the connection
+	// when response handling is completed.
+	w.Header().Set("Connection", "close")
+	w.Header().Set("Content-Type", "application/json")
+	r.Close = true
+
+	body, err := ioutil.ReadAll(r.Body)
+	r.Body.Close()
+	if err != nil {
+		http.Error(w, "error reading request body", http.StatusBadRequest)
+		return
+	}
+	req := new(msgjson.Message)
+	err = json.Unmarshal(body, req)
+	if err != nil {
+		http.Error(w, "JSON decode error", http.StatusUnprocessableEntity)
+		return
+	}
+	if req.Type != msgjson.Request {
+		http.Error(w, "Responses not accepted", http.StatusMethodNotAllowed)
+		return
+	}
+	s.parseHTTPRequest(w, req)
+}
+
+// Config holds variables neede to create a new RPC Server.
+type Config struct {
+	Core                        ClientCore
+	Addr, User, Pass, Cert, Key string
+}
+
+// SetLogger sets the logger for the RPCServer package.
+func SetLogger(logger slog.Logger) {
 	log = logger
-	log.Infof("RPC server running at %s", addr)
-	<-ctx.Done()
+}
+
+// New is the constructor for an RPCServer.
+func New(cfg *Config) (*RPCServer, error) {
+
+	// Find or create the key pair.
+	keyExists := fileExists(cfg.Key)
+	certExists := fileExists(cfg.Cert)
+	if certExists == !keyExists {
+		return nil, fmt.Errorf("missing cert pair file")
+	}
+	if !keyExists && !certExists {
+		err := genCertPair(cfg.Cert, cfg.Key)
+		if err != nil {
+			return nil, err
+		}
+	}
+	keypair, err := tls.LoadX509KeyPair(cfg.Cert, cfg.Key)
+	if err != nil {
+		return nil, err
+	}
+
+	// Prepare the TLS configuration.
+	tlsConfig := tls.Config{
+		Certificates: []tls.Certificate{keypair},
+		MinVersion:   tls.VersionTLS12,
+	}
+
+	// Create listener.
+	listener, err := tls.Listen("tcp", cfg.Addr, &tlsConfig)
+	if err != nil {
+		return nil, fmt.Errorf("Can't listen on %s. web server quitting: %v", cfg.Addr, err)
+	}
+
+	// Create an HTTP router.
+	mux := chi.NewRouter()
+	httpServer := &http.Server{
+		Handler:      mux,
+		ReadTimeout:  rpcTimeoutSeconds * time.Second, // slow requests should not hold connections opened
+		WriteTimeout: rpcTimeoutSeconds * time.Second, // hung responses must die
+	}
+
+	// Make the server.
+	s := &RPCServer{
+		core:     cfg.Core,
+		listener: listener,
+		srv:      httpServer,
+		syncers:  make(map[string]*marketSyncer),
+		clients:  make(map[int32]*wsClient),
+	}
+
+	// Create authsha to verify requests against.
+	if cfg.User != "" && cfg.Pass != "" {
+		login := cfg.User + ":" + cfg.Pass
+		auth := "Basic " +
+			base64.StdEncoding.EncodeToString([]byte(login))
+		s.authsha = sha256.Sum256([]byte(auth))
+	}
```

Comment:
You don't appear to be covering auth in testing. Setting the user and password should get coverage here and `authMiddleware`, right?